src/compile/data/stack.ts

```diff
@@ -211,58 +204,54 @@ export class StackNode extends DataFlowNode {
 
   public assemble(): VgTransform[] {
     const transform: VgTransform[] = [];
-
-    // Detects whether assemble call is from Node created from make or MakeFromTransform
-    // TODO Create a stricter function to check that.
-    if (this._stack.stackby) {
-      const {facetby, field: stackField, dimensionFieldDef, impute, offset, sort, stackby} = this._stack;
+    const {facetby, dimensionFieldDef, stackField: field, stackby, sort, offset, impute, groupby, as} = this._stack;
 
       // Impute
-      if (impute && dimensionFieldDef) {
-        const dimensionField = dimensionFieldDef ? vgField(dimensionFieldDef, {binSuffix: 'mid'}): undefined;
-
-        if (dimensionFieldDef.bin) {
-          // As we can only impute one field at a time, we need to calculate
-          // mid point for a binned field
-          transform.push({
-            type: 'formula',
-            expr: '(' +
-              vgField(dimensionFieldDef, {expr: 'datum'}) +
-              '+' +
-              vgField(dimensionFieldDef, {expr: 'datum', binSuffix: 'end'}) +
-              ')/2',
-            as: dimensionField
-          });
-        }
+    if (impute && dimensionFieldDef) {
+      const dimensionField = dimensionFieldDef ? vgField(dimensionFieldDef, {binSuffix: 'mid'}): undefined;
 
+      if (dimensionFieldDef.bin) {
+        // As we can only impute one field at a time, we need to calculate
+        // mid point for a binned field
         transform.push({
-          type: 'impute',
-          field: stackField,
-          groupby: stackby,
-          key: dimensionField,
-          method: 'value',
-          value: 0
+          type: 'formula',
+          expr: '(' +
+            vgField(dimensionFieldDef, {expr: 'datum'}) +
+            '+' +
+            vgField(dimensionFieldDef, {expr: 'datum', binSuffix: 'end'}) +
+            ')/2',
+          as: dimensionField
         });
       }
 
+      transform.push({
+        type: 'impute',
+        field,
+        groupby: stackby,
+        key: dimensionField,
+        method: 'value',
+        value: 0
+      });
+    }
+    if(facetby) {
```

Comment:
Why adding this `if(facetby) {`?

Reply:
I'm not sure about this since I haven't read how `facetby` is generated, but the general logic behind the `if` is that there are 3 code segments in assemble: 
* Segment 1: when `this._stack` has been created by `makeFromEncoding` where `impute`: `true`
* Segment 2: when `this._stack` has been created by `makeFromEncoding`
* Segment 3. when `this._stack` has been created by `makeFromTransform`

The existing code structure led me to believe that Segment 1 and Segment 2 are not disjoint and both can be executed for a single input spec.

To maintain a flat structure, initially `this._stack` with `impute:true` is handled, then all the `this._stack` created by `makeFromEncoding`(**including** the ones operated on in the previous step), then finally the ones created by `makeFromTransform` which is disjoint to Segment 1 and 2.

`facetby` is `undefined` for any `this._stack` created by `makeFromTransform` . So those cases skip the `if(facetby) {` block.